cpp/src/quantiles/quantiles.cu

```diff
@@ -16,196 +16,253 @@
-                         void* t_erased_res,
-                         gdf_context* ctxt)
-  {
-    ColType* ptr_res = static_cast<ColType*>(t_erased_res);
-    size_t n = col_in->size;
-    ColType* p_dv = static_cast<ColType*>(col_in->data);
-    if( ctxt->flag_sort_inplace || ctxt->flag_sorted )
-      {
-        *ptr_res = quantile_approx(p_dv, n, q, NULL, ctxt->flag_sorted);
-      }
-    else
-      {
-        rmm::device_vector<ColType> dv(n);
-        thrust::copy_n(thrust::device, /*TODO: stream*/p_dv, n, dv.begin());
-        cudaDeviceSynchronize();
-        p_dv = dv.data().get();
-
-        *ptr_res = quantile_approx(p_dv, n, q, NULL, ctxt->flag_sorted);
+                               ctxt->flag_sorted, stream);
       }
   }
     
   struct trampoline_exact_functor{
     template <typename T,
               typename std::enable_if_t<!std::is_arithmetic<T>::value, int> = 0>
     gdf_error operator()(gdf_column* col_in,
-                         gdf_quantile_method prec,
-                         double q,
-                         void* t_erased_res,
-                         gdf_context* ctxt)
+                         gdf_quantile_method interpolation,
+                         double              quant,
+                         void*               t_erased_res,
+                         gdf_context*        ctxt,
+                         cudaStream_t        stream = NULL)
     {
       return GDF_UNSUPPORTED_DTYPE;
     }
 
     template <typename T,
               typename std::enable_if_t<std::is_arithmetic<T>::value, int> = 0>
-    gdf_error operator()(gdf_column*  col_in,              //input column;
-                         gdf_quantile_method prec,         //precision: type of quantile method calculation
-                         double              q,            //requested quantile in [0,1]
-                         void*               t_erased_res, //result; for <exact> should probably be double*; it's void* because
-                                                           //(1) for uniformity of interface with <approx>;
-                                                           //(2) for possible types bigger than double, in the future;
-                         gdf_context*        ctxt)         //context info
+    gdf_error operator()(gdf_column*  col_in,
+                         gdf_quantile_method interpolation,
+                         double              quant,
+                         void*               t_erased_res,
+                         gdf_context*        ctxt,
+                         cudaStream_t        stream = NULL)
     {
-      return trampoline_exact<T>(col_in, prec, q, t_erased_res, ctxt);
+      // just in case double won't be enough to hold result
+      // it can be changed in future
+      return trampoline_exact<T, double>
+                 (col_in, interpolation, quant, t_erased_res, ctxt, stream);
     }
   };
 
   struct trampoline_approx_functor{
     template <typename T>

```

Comment:
trampoline_exact_functor has enable_if, but the approx_functor doesn't. Is that correct?